cajita/src/Cajita_Splines.hpp

```diff
@@ -446,72 +448,321 @@ struct Spline<3>
 
-    // Weight gradients.
+    // Weight physical gradients.
     Scalar g[3][num_knot];
 };
 
+// Spline data container.
+template <typename Scalar, int Order, class EntityType, class Tags = void>
+struct SplineData;
+
+// Default of void has all data members.
+template <typename Scalar, int Order, class EntityType>
+struct SplineData<Scalar, Order, EntityType, void>
+    : SplineDataMember<Scalar, Order, SplinePhysicalCellSize>,
+      SplineDataMember<Scalar, Order, SplineLogicalPosition>,
+      SplineDataMember<Scalar, Order, SplinePhysicalDistance>,
+      SplineDataMember<Scalar, Order, SplineWeightValues>,
+      SplineDataMember<Scalar, Order, SplineWeightPhysicalGradients>
+{
+    using scalar_type = Scalar;
+    static constexpr int order = Order;
+    using spline_type = Spline<Order>;
+    static constexpr int num_knot = spline_type::num_knot;
+    using entity_type = EntityType;
+
+    using has_physical_cell_size = std::true_type;
+    using has_logical_position = std::true_type;
+    using has_physical_distance = std::true_type;
+    using has_weight_values = std::true_type;
+    using has_weight_physical_gradients = std::true_type;
+
+    // Local interpolation stencil.
+    int s[3][num_knot];
+};
+
+// Specify each data member individually through tags.
+template <typename Scalar, int Order, class EntityType, class... Tags>
+struct SplineData<Scalar, Order, EntityType, SplineDataMemberTypes<Tags...>>
+    : SplineDataMember<Scalar, Order, Tags>...
+{
+    using scalar_type = Scalar;
+    static constexpr int order = Order;
+    using spline_type = Spline<Order>;
+    static constexpr int num_knot = spline_type::num_knot;
+    using entity_type = EntityType;
+
+    using member_tags = SplineDataMemberTypes<Tags...>;
+
+    using has_physical_cell_size =
+        has_spline_tag<SplinePhysicalCellSize, member_tags>;
+    using has_logical_position =
+        has_spline_tag<SplineLogicalPosition, member_tags>;
+    using has_physical_distance =
+        has_spline_tag<SplinePhysicalDistance, member_tags>;
+    using has_weight_values = has_spline_tag<SplineWeightValues, member_tags>;
+    using has_weight_physical_gradients =
+        has_spline_tag<SplineWeightPhysicalGradients, member_tags>;
+
+    // Local interpolation stencil.
+    int s[3][num_knot];
+};
+
+// Assign physical cell size to the spline data.
+template <typename Scalar, int Order, class EntityType, class DataTags>
+KOKKOS_INLINE_FUNCTION void
+setSplineData( std::true_type, SplinePhysicalCellSize,
+               SplineData<Scalar, Order, EntityType, DataTags> &data,
+               const int d, const Scalar dx )
+{
+    data.dx[d] = dx;
+}
+
+template <typename Scalar, int Order, class EntityType, class DataTags>
+KOKKOS_INLINE_FUNCTION void
+setSplineData( std::false_type, SplinePhysicalCellSize,
+               SplineData<Scalar, Order, EntityType, DataTags> &, const int,
+               const Scalar )
+{
+}
```

Comment:
Why do you make the caller pass `true_type` or `false_type`?
Why don't you just use SFINAE?
```suggestion
// Assign physical cell size to the spline data.
template <typename Scalar, int Order, class EntityType, class DataTags>
KOKKOS_INLINE_FUNCTION std::enable_if_t<typename SplineData<Scalar, Order, EntityType, DataTags>::has_physical_cell_size{}>
setSplineData( SplinePhysicalCellSize,
               SplineData<Scalar, Order, EntityType, DataTags> &data,
               const int d, const Scalar dx )
{
    data.dx[d] = dx;
}

template <typename Scalar, int Order, class EntityType, class DataTags>
KOKKOS_INLINE_FUNCTION std::enable_if_t<!typename SplineData<Scalar, Order, EntityType, DataTags>::has_physical_cell_size{}>
setSplineData( SplinePhysicalCellSize,
               SplineData<Scalar, Order, EntityType, DataTags> &, const int,
               const Scalar )
{
}
```

Reply:
good idea - I implemented this